Source/CarthageKit/Algorithms.swift

```diff
@@ -65,3 +65,38 @@ public func topologicalSort<Node: Comparable>(graph: Dictionary<Node, Set<Node>>
 
 	return workingGraph.isEmpty ? sorted : nil
 }
+
+/// Performs a topological sort on the provided graph with its output sorted to
+/// include only the provided set of nodes and their transitively incoming 
+/// nodes (dependencies).
+///
+/// Returns nil if any of the provided nodes are not contained within the
+/// provided graph or if the provided graph has a cycle.
+public func topologicalSort<Node: Comparable>(graph: Dictionary<Node, Set<Node>>, nodes: Set<Node>) -> [Node]? {
+	guard !nodes.isEmpty else { return topologicalSort(graph) }
+
+	guard nodes.isSubsetOf(Set(graph.keys)) else { return nil }
+
+	// Ensure that the graph has no cycles, otherwise determining the set of 
+	// transitive incoming nodes could infinitely recurse.
+	guard let _ = topologicalSort(graph) else { return nil }
+
+	let relevantNodes = Set(nodes.flatMap { Set([$0]).union(transitiveIncomingNodes(graph, node: $0)) })
+	let irrelevantNodes = Set(graph.keys).subtract(relevantNodes)
+
+	var filteredGraph = graph
+	irrelevantNodes.forEach { node in filteredGraph.removeValueForKey(node) }
+	
+	return topologicalSort(filteredGraph)
```

Comment:
It doesn't seem like we should need to sort again. Can't we take the result of the earlier call to `topologicalSort` and remove the `irrelaventNodes` from the result?